src/sigma/sigma_primitives.hpp

```diff
@@ -71,34 +63,33 @@ std::vector<uint64_t> SigmaPrimitives<Exponent, GroupElement>::convert_to_nal(
 }
 
 template<class Exponent, class GroupElement>
-void SigmaPrimitives<Exponent, GroupElement>::get_x(
-        const GroupElement& A,
-        const GroupElement& C,
-        const GroupElement& D,
+void SigmaPrimitives<Exponent, GroupElement>::generate_challenge(
+        const std::vector<GroupElement>& group_elements,
         Exponent& result_out) {
     CSHA256 hash;
-    unsigned char data[3 * A.serialize_size];
-    unsigned char* current = A.serialize(data);
-    current = C.serialize(current);
-    D.serialize(current);
-    hash.Write(data, 3 * A.memoryRequired());
+    std::vector<unsigned char> data(group_elements.size() * group_elements[0].memoryRequired());
+    unsigned char* current = data.data();
+    for (size_t i = 0; i < group_elements.size(); ++i) {
+        current = group_elements[i].serialize(current);
+    }
+    hash.Write(data.data(), data.size());
```

Comment:
IIRC `hash.Write` will calculate a hash of data based of previous write. That mean we don't need to concatenate into `data`  by ourself.